File: static/CWE-611/csharp/patched.cs
// Remediation for CWE-611: Improper Restriction of XML External Entity Reference ('XXE')
// Fix: Disable DTD processing and external resolution via XmlReaderSettings and a null XmlResolver.

using System.Xml;

namespace GuineaPig.Sast
{
    class Patched_611
    {
        public string ParseXml(string userInput)
        {
            var settings = new XmlReaderSettings
            {
                DtdProcessing = DtdProcessing.Prohibit,
                XmlResolver = null
            };
            var doc = new XmlDocument { XmlResolver = null };
            using (var reader = XmlReader.Create(new System.IO.StringReader(userInput), settings))
            {
                doc.Load(reader);
            }
            return doc.OuterXml;
        }
    }
}

File: static/CWE-611/java/patched.java
// Remediation for CWE-611: Improper Restriction of XML External Entity Reference (XXE)
// Fix: Enable secure processing and disallow DOCTYPE declarations and external entities.

import java.io.ByteArrayInputStream;
import javax.xml.XMLConstants;
import javax.xml.parsers.DocumentBuilder;
import javax.xml.parsers.DocumentBuilderFactory;
import org.w3c.dom.Document;

class Patched_611 {
    public Document parseXml(String userInput) throws Exception {
        DocumentBuilderFactory dbf = DocumentBuilderFactory.newInstance();
        dbf.setFeature(XMLConstants.FEATURE_SECURE_PROCESSING, true);
        dbf.setFeature("http://apache.org/xml/features/disallow-doctype-decl", true);
        dbf.setFeature("http://xml.org/sax/features/external-general-entities", false);
        dbf.setFeature("http://xml.org/sax/features/external-parameter-entities", false);
        dbf.setExpandEntityReferences(false);
        DocumentBuilder db = dbf.newDocumentBuilder();
        return db.parse(new ByteArrayInputStream(userInput.getBytes("UTF-8")));
    }
}
